envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/snapshot/listeners/filters/RBACFilterFactory.kt

```diff
@@ -163,22 +176,80 @@ class RBACFilterFactory(
         }
     }
 
-    private fun sourceIpPrincipals(client: String, snapshot: GlobalSnapshot): List<Principal> {
-        val clientEndpoints = snapshot.endpoints.resources().filterKeys { client == it }.values
+    private fun ipFromDiscoveryPrincipals(
+        decomposed: String,
+        selector: Selector?,
+        matching: Matching?,
+        snapshot: GlobalSnapshot
+    ): List<Principal> {
+        val clientEndpoints = snapshot.endpoints.resources().filterKeys { decomposed == it }.values
         return clientEndpoints.flatMap { clusterLoadAssignment ->
             clusterLoadAssignment.endpointsList.flatMap { lbEndpoints ->
                 lbEndpoints.lbEndpointsList.map { lbEndpoint ->
                     lbEndpoint.endpoint.address
                 }
             }
-        }.map { address ->
-            Principal.newBuilder().setSourceIp(CidrRange.newBuilder()
+        }.flatMap { address ->
+            val sourceIpPrincipal = Principal.newBuilder()
+                    .setSourceIp(CidrRange.newBuilder()
                     .setAddressPrefix(address.socketAddress.address)
                     .setPrefixLen(EXACT_IP_MASK).build())
+
+            if (selector != null && matching != null) {
+                addAdditionalAuthentication(selector, matching, listOf(sourceIpPrincipal))
+            } else {
+                listOf(sourceIpPrincipal.build())
+            }
+        }
+    }
+
+    private fun addAdditionalAuthentication(
+        selector: Selector,
+        matching: Matching,
+        sourceIpPrincipals: List<Principal.Builder>
+    ): List<Principal> {
+        val andPrincipal = Principal.newBuilder()
+        val (matchingType, matchingValue) = matching
+
+        val additionalAuthenticationPrincipal = when (matchingType) {
+            AdditionalAuthenticationMethod.HEADER -> Principal.newBuilder()
+                    .setHeader(
+                        HeaderMatcher.newBuilder()
+                            .setName(matchingValue)
+                            .setExactMatch(selector)
+                    )
+        }
+
+        return sourceIpPrincipals.map {
+            val principalSet = Principal.Set.newBuilder()
+                    .addAllIds(listOf(it.build(), additionalAuthenticationPrincipal.build()))
                     .build()
+
+            andPrincipal.setAndIds(principalSet).build()
         }
     }
 
+    private fun decompose(client: String): Triple<String, Selector?, Matching?> {
```

Comment:
This tells me that `client` deserves better abstraction than String :)

Reply:
This is the raw client string that comes from metadata (`IncomingEndpoint.client`), please share your ideas on naming / typealias / structure.

Reply:
I would create a dedicated class for it, however, since I haven't looked deeply at this PR I have not idea how labor-intensive it would be.